File: GlueCustomConnectors/development/Spark/glue-3.0/tpcds-custom-connector-for-glue3.0/jobvalidation/scala/GlueJobValidationDataSchemaTest.scala
/*
 * Copyright 2016-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 */

import com.amazonaws.services.glue.GlueContext
import com.amazonaws.services.glue.errors.CallSite
import com.amazonaws.services.glue.schema.{Schema, TypeCode}
import com.amazonaws.services.glue.schema.builders.SchemaBuilder
import com.amazonaws.services.glue.schema.types.DecimalType
import com.amazonaws.services.glue.util.GlueArgParser
import com.amazonaws.services.glue.util.Job
import com.amazonaws.services.glue.util.JsonOptions
import org.apache.spark.SparkContext

import scala.collection.JavaConverters._

object GlueJobValidationDataSchemaTest {
  def main(sysArgs: Array[String]) {
    val spark: SparkContext = new SparkContext()
    val glueContext: GlueContext = new GlueContext(spark)

    val optionsMap = Map(
      "table" -> "call_center",
      "scale" -> "1",
      "numPartitions" -> "1",
      "connectionName" -> "GlueTPCDSConnection"
    )

    // create DataSource and read data
    val customSource = glueContext.getSource(
      connectionType = "marketplace.spark",
      connectionOptions = JsonOptions(optionsMap),
      transformationContext = "customSource")
    val dyf = customSource.getDynamicFrame()

    // verify schema of 'customer' table
    val expectedSchema = new Schema(new SchemaBuilder()
      .beginStruct()
      .atomicField("cc_call_center_sk", TypeCode.LONG)
      .atomicField("cc_call_center_id", TypeCode.STRING)
      .atomicField("cc_rec_start_date", TypeCode.DATE)
      .atomicField("cc_rec_end_date", TypeCode.DATE)
      .atomicField("cc_closed_date_sk", TypeCode.INT)
      .atomicField("cc_open_date_sk", TypeCode.INT)
      .atomicField("cc_name", TypeCode.STRING)
      .atomicField("cc_class", TypeCode.STRING)
      .atomicField("cc_employees", TypeCode.INT)
      .atomicField("cc_sq_ft", TypeCode.INT)
      .atomicField("cc_hours", TypeCode.STRING)
      .atomicField("cc_manager", TypeCode.STRING)
      .atomicField("cc_mkt_id", TypeCode.INT)
      .atomicField("cc_mkt_class", TypeCode.STRING)
      .atomicField("cc_mkt_desc", TypeCode.STRING)
      .atomicField("cc_market_manager", TypeCode.STRING)
      .atomicField("cc_division", TypeCode.INT)
      .atomicField("cc_division_name", TypeCode.STRING)
      .atomicField("cc_company", TypeCode.INT)
      .atomicField("cc_company_name", TypeCode.STRING)
      .atomicField("cc_street_number", TypeCode.STRING)
      .atomicField("cc_street_name", TypeCode.STRING)
      .atomicField("cc_street_type", TypeCode.STRING)
      .atomicField("cc_suite_number", TypeCode.STRING)
      .atomicField("cc_city", TypeCode.STRING)
      .atomicField("cc_county", TypeCode.STRING)
      .atomicField("cc_state", TypeCode.STRING)
      .atomicField("cc_zip", TypeCode.STRING)
      .atomicField("cc_country", TypeCode.STRING)
      .atomicField("cc_gmt_offset", new DecimalType(5, 2))
      .atomicField("cc_tax_percentage", new DecimalType(5, 2))
      .endStruct().build())

    assert(dyf.schema == expectedSchema)
  }
}

File: GlueCustomConnectors/development/Spark/glue-3.0/tpcds-custom-connector-for-glue3.0/jobvalidation/pyspark/glue_job_validation_data_schema_test.py
#  Copyright 2016-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#  SPDX-License-Identifier: MIT-0

from pyspark.context import SparkContext
from awsglue.context import GlueContext
from awsglue.gluetypes import Field, IntegerType, LongType, StructType, StringType, DecimalType, DateType


glue_context = GlueContext(SparkContext())
logger = glue_context.get_logger()

connection_options = {
    "table": "call_center",
    "scale": "1",
    "numPartitions": "1",
    "connectionName" : "GlueTPCDSConnection"
}

# read data from data source
datasource0 = glue_context.create_dynamic_frame_from_options(
    connection_type="marketplace.spark", # use `custom.spark` when running job validation tests by self-build custom connector
    connection_options=connection_options)

# validate data schema
expected_schema = StructType([
    Field("cc_call_center_sk", LongType()),
    Field("cc_call_center_id", StringType()),
    Field("cc_rec_start_date", DateType()),
    Field("cc_rec_end_date", DateType()),
    Field("cc_closed_date_sk", IntegerType()),
    Field("cc_open_date_sk", IntegerType()),
    Field("cc_name", StringType()),
    Field("cc_class", StringType()),
    Field("cc_employees", IntegerType()),
    Field("cc_sq_ft", IntegerType()),
    Field("cc_hours", StringType()),
    Field("cc_manager", StringType()),
    Field("cc_mkt_id", IntegerType()),
    Field("cc_mkt_class", StringType()),
    Field("cc_mkt_desc", StringType()),
    Field("cc_market_manager", StringType()),
    Field("cc_division", IntegerType()),
    Field("cc_division_name", StringType()),
    Field("cc_company", IntegerType()),
    Field("cc_company_name", StringType()),
    Field("cc_street_number", StringType()),
    Field("cc_street_name", StringType()),
    Field("cc_street_type", StringType()),
    Field("cc_suite_number", StringType()),
    Field("cc_city", StringType()),
    Field("cc_county", StringType()),
    Field("cc_state", StringType()),
    Field("cc_zip", StringType()),
    Field("cc_country", StringType()),
    Field("cc_gmt_offset", DecimalType(precision=5, scale=2)),
    Field("cc_tax_percentage", DecimalType(precision=5, scale=2))
])

result_schema = datasource0.schema()
assert result_schema == expected_schema
logger.info(f'Expected schema: {expected_schema.jsonValue()}')
logger.info(f'Result schema: {result_schema.jsonValue()}')
logger.info("Result schema in tree structure: ")
datasource0.printSchema()